pkg/controller/zerocapacity/zero_controller.go

```diff
@@ -0,0 +1,467 @@
+		if errors.IsNotFound(err) {
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if err := ensureClusterStability(infinispan); err != nil {
+		return reconcile.Result{}, fmt.Errorf("Infinispan not stable: %w", err)
+	}
+
+	statefulset := &appsv1.StatefulSet{}
+	if err := z.Client.Get(ctx, clusterKey, statefulset); err != nil {
+		z.Log.Info(fmt.Sprintf("Unable to load Infinispan StatefulSet '%s': %w", clusterName, err))
+		if errors.IsNotFound(err) {
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	spec, err := instance.Init()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	configMap, err := z.configureZeroCapacity(name, namespace, spec, infinispan, instance)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Unable to create zero-capacity configuration: %w", err)
+	}
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels:    spec.PodLabels,
+		},
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, z.Client, pod, func() error {
+		z.configureZeroPod(name, configMap, infinispan, spec, pod, statefulset.Spec.Template.Spec)
+		return controllerutil.SetControllerReference(instance.AsMeta(), pod, z.Scheme)
+	})
+
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Unable to CreateOrUpdate zero-capacity pod: %w", err)
+	}
+
+	// Update status
+	return reconcile.Result{}, instance.UpdatePhase(ZeroInitialized, "")
+}
+
+func ensureClusterStability(infinispan *v1.Infinispan) error {
+	conditions := map[string]metav1.ConditionStatus{
+		v1.ConditionGracefulShutdown:   metav1.ConditionFalse,
+		v1.ConditionPrelimChecksFailed: metav1.ConditionFalse,
+		v1.ConditionStopping:           metav1.ConditionFalse,
+		v1.ConditionWellFormed:         metav1.ConditionTrue,
+	}
+	return infinispan.ExpectConditionStatus(conditions, true)
+}
+
+func (z *Controller) execute(httpClient http.HttpClient, request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	if !z.isZeroPodReady(request, instance) {
+		// Don't requeue as reconcile request is received when the zero pod becomes ready
+		return reconcile.Result{}, nil
+	}
+
+	if err := instance.Exec(httpClient); err != nil {
+		z.Log.Error(err, "Unable to execute action on zero-capacity pod", "request.Name", request.Name)
+		return reconcile.Result{}, instance.UpdatePhase(ZeroFailed, err.Error())
+	}
+
+	return reconcile.Result{}, instance.UpdatePhase(ZeroRunning, "")
+}
+
+func (z *Controller) waitForExecutionToComplete(httpClient http.HttpClient, request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	phase, err := instance.ExecStatus(httpClient)
+
+	if err != nil || phase == ZeroFailed {
+		z.Log.Error(err, "Execution failed", "request.Name", request.Name)
+		return reconcile.Result{}, instance.UpdatePhase(ZeroFailed, err.Error())
```

Comment:
@ryanemerson this can fails with `panic` as `err` could be `nil` you try to return `err.Error()`